hal/armv7a/zynq7000/zynq.c

```diff
@@ -91,6 +90,331 @@ static void _zynq_slcrUnlock(void)
+
+		case pctl_ctrl_can_clk:
+			val = *(zynq_common.slcr + slcr_can_clk_ctrl);
+			*clkact0 = val & 0x1;
+			*clkact1 = (val >> 1) & 0x1;
+			*srcsel = (val >> 4) & 0x3;
+			*divisor0 = (val >> 8) & 0x3f;
+			*divisor1 = (val >> 20) & 0x3f;
+			break;
+
+		default:
+			return -1;
+	}
+
+	return 0;
+}
+
+
+static int _zynq_setMioClk(int mio, char ref0, char mux0, char ref1, char mux1)
+{
+	if (mio == pctl_ctrl_can_mioclk) {
+		_zynq_slcrUnlock();
+		*(zynq_common.slcr + slcr_can_mioclk_ctrl) = (*(zynq_common.slcr + slcr_can_mioclk_ctrl) & ~0x007f007f) | (mux0 & 0x3f) | ((!!ref0) << 6) |
+			((mux1 & 0x3f) << 16) | ((!!ref1) << 22);
+		_zynq_slcrLock();
+
+		return 0;
+	}
+
+	return -1;
+}
+
+
+static int _zynq_getMioClk(int mio, char *ref0, char *mux0, char *ref1, char *mux1)
+{
+	u32 val = 0;
+
+	if (mio == pctl_ctrl_can_mioclk) {
+		val = *(zynq_common.slcr + slcr_can_mioclk_ctrl);
+		*mux0 = val & 0x3f;
+		*ref0 = (val >> 6) & 0x1;
+		*mux1 = (val >> 16) & 0x3f;
+		*ref1 = (val >> 22) & 0x1;
+
+		return 0;
+	}
+
+	return -1;
+}
+
+
+static int _zynq_setMIO(int pin, char disableRcvr, char pullup, char ioType, char speed, char l0, char l1, char l2, char l3, char triEnable)
+{
+	u32 val = 0;
+
+	if (pin > 53)
+		return -1;
+
+	val = (!!triEnable) | (!!l0 << 1) | (!!l1 << 2) | ((l2 & 0x3) << 3) |
+		((l3 & 0x7) << 5) | (!!speed << 8) | ((ioType & 0x7) << 9) | (!!pullup << 12) |
+		(!!disableRcvr << 13);
+
+	_zynq_slcrUnlock();
+	*(zynq_common.slcr + slcr_mio_pin_00 + pin) = (*(zynq_common.slcr + slcr_mio_pin_00 + pin) & ~0x00003fff) | val;
+	_zynq_slcrLock();
+
+	return 0;
+}
+
+
+static int _zynq_getMIO(int pin, char *disableRcvr, char *pullup, char *ioType, char *speed, char *l0, char *l1, char *l2, char *l3, char *triEnable)
+{
+	u32 val;
+
+	if (pin > 53)
+		return -1;
+
+	val = *(zynq_common.slcr + slcr_mio_pin_00 + pin);
+
+	*disableRcvr = (val >> 13) & 0x1;
+	*pullup = (val >> 12) & 0x1;
+	*ioType = (val >> 9) & 0x7;
+	*speed = (val >> 8) & 0x1;
+	*l0 = (val >> 1) & 0x1;
+	*l1 = (val >> 2) & 0x1;
+	*l2 = (val >> 3) & 0x3;
+	*l3 = (val >> 5) & 0x7;
+	*triEnable = val & 0x1;
+
+	return 0;
+}
+
+
+int hal_platformctl(void *ptr)
+{
+	platformctl_t *data = ptr;
+	unsigned int t = 0;
+	spinlock_ctx_t sc;
+	int ret = -1;
+
+	hal_spinlockSet(&zynq_common.pltctlSp, &sc);
+
+	switch (data->type) {
+		case pctl_ambaclock:
+			if (data->action == pctl_set) {
+				ret = _zynq_setAmbaClk(data->ambaclock.dev, data->ambaclock.state);
+			}
+			else if (data->action == pctl_get) {
+				ret = _zynq_getAmbaClk(data->ambaclock.dev, &t);
+				data->ambaclock.state = t;
+			}
+			break;
+
+		case pctl_mioclock:
+			if (data->action == pctl_set)
+				ret = _zynq_setMioClk(data->mioclock.mio, data->mioclock.ref0,
+					data->mioclock.mux0, data->mioclock.ref1, data->mioclock.mux1);
+			else if (data->action == pctl_get)
+				ret = _zynq_getMioClk(data->mioclock.mio, &data->mioclock.ref0,
+					&data->mioclock.mux0, &data->mioclock.ref1, &data->mioclock.mux1);
```

Comment:
platformctl_t should be packed - so you're taking address of packed struct member here

Reply:
It's ok, it's a char - pointer can't be misaligned